lib/src/lib.rs

```diff
@@ -999,21 +1000,41 @@ impl Repo {
     }
 
     fn chunk_writer(&self, rx: mpsc::Receiver<ChunkMessage>) {
+        let mut queue: VecDeque<(PathBuf, PathBuf, File)> = VecDeque::with_capacity(CHANNEL_SIZE);
+        // Cache paths to make sure we don't queue up an existing block
+        let mut queue_paths: HashSet<PathBuf> = HashSet::new();
+
+        fn flush(queue: &VecDeque<(PathBuf, PathBuf, File)>) {
+            for &(_, _, ref chunk_file) in queue.iter() {
+                chunk_file.sync_data().unwrap();
+                drop(chunk_file);
+            }
+            // Run renames after data sync to mitigate data writes and metadata writes contention
+            for &(ref path, ref tmp_path, _) in queue.iter() {
+                fs::rename(&tmp_path, &path).unwrap();
+            }
+        }
         loop {
             match rx.recv().unwrap() {
                 ChunkMessage::Exit => {
+                    flush(&queue);
                     return;
                 }
                 ChunkMessage::Data(data, sha256, chunk_type, _) => {
                     let path = self.chunk_path_by_digest(&sha256, chunk_type);
-                    if !path.exists() {
+                    if !queue_paths.contains(&path) && !path.exists() {
                         let tmp_path = path.with_extension("tmp");
                         fs::create_dir_all(path.parent().unwrap()).unwrap();
                         let mut chunk_file = fs::File::create(&tmp_path).unwrap();
                         chunk_file.write_all(&data).unwrap();
-                        chunk_file.sync_data().unwrap();
-                        drop(chunk_file);
-                        fs::rename(&tmp_path, &path).unwrap();
+
+                        // Queue chunk up for data sync and rename
+                        queue_paths.insert(path.clone());
```

Comment:
Well technically we currently are since we create the index as one big in memory vector but that would reduce the max size of the index by half before we run out of memory. We can explore this in a separate issue. I have changed the code to drain queue_paths along side the queue so it will only ever be the same size as the queue and never more which is currently 128.